src/hammer-vlsi/synthesis/yosys/__init__.py

```diff
@@ -0,0 +1,253 @@
+
+            if new_path == "":
+                assert constraint.type == PlacementConstraintType.TopLevel, \
+                    "Top must be a top-level/chip size constraint"
+                margins = constraint.margins
+                assert margins is not None
+                # Set top-level chip dimensions.
+                return "{} {} {} {}".format(
+                    constraint.x, constraint.y, 
+                    constraint.width, constraint.height)
+
+        raise Exception("no top-level placement constraint was found")
+
+
+    #========================================================================
+    # synthesis steps below
+    #========================================================================
+
+    def setup_env(self) -> bool:
+        """OPENROAD must be sucessfully installed before using it"""
+
+        if "OPENROAD" not in os.environ:
+            raise Exception("OPENROAD is not defined in environment!")
+        if not shutil.which("klayout"):
+            raise Exception("klayout is not in PATH")
+
+        # TODO: for now, just symlink in the read-only OpenROAD stuff, since
+        # the $OPENROAD/flow/Makefile expects these in the current directory.
+        # in the future, $OPENROAD/flow/Makefile should use ?= instead of =
+        openroad = self.openroad_flow_path()
+        for subpath in ["platforms", "scripts", "util", "test"]:
+          src = os.path.join(openroad, "flow/{}".format(subpath))
+          dst = os.path.join(self.run_dir, subpath)
+
+          try:
+            os.remove(dst)
+          except:
+            pass
+
+          os.symlink(src, dst)
+
+        return True
+
+    def create_synth_script(self) -> bool:
+        """right now, just using the OpenROAD built-in synth script"""
+        return True
+
+    def create_design_config(self) -> bool:
+        design_config = self.design_config_path()
+
+        # Load input files and check that they are all Verilog.
+        if not self.check_input_files([".v", ".sv"]):
+            return False
+        abspath_input_files = list(map(lambda name: 
+          os.path.join(os.getcwd(), name), self.input_files))
+
+        # Add any verilog_synth wrappers (which are needed in some 
+        # technologies e.g. for SRAMs) which need to be synthesized.
+        abspath_input_files += self.technology.read_libs([
+            hammer_tech.filters.verilog_synth_filter
+        ], hammer_tech.HammerTechnologyUtils.to_plain_item)
+
+        # Generate constraints
+        input_sdc = os.path.join(self.run_dir, "input.sdc")
+        unit = self.get_time_unit().value_prefix + self.get_time_unit().unit
+        with open(input_sdc, "w") as f:
+            f.write("set_units -time {}\n".format(unit))
+            f.write(self.sdc_clock_constraints)
+            f.write("\n")
+            f.write(self.sdc_pin_constraints)
+
+        # TODO: i am blindly reading in all libs for all corners.
```

Comment:
Does that hurt runtime?

Reply:
not for nangate45, or anything else small enough to use openroad